src/mlpack/tests/rbm_network_test.cpp

```diff
@@ -0,0 +1,158 @@
+#include <mlpack/methods/ann/rbm.hpp>
+#include <mlpack/methods/softmax_regression/softmax_regression.hpp>
+#include <mlpack/core/optimizers/minibatch_sgd/minibatch_sgd.hpp>
+#include <mlpack/core/optimizers/sgd/sgd.hpp>
+#include <mlpack/core/optimizers/lbfgs/lbfgs.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "test_tools.hpp"
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+using namespace mlpack::regression;
+
+BOOST_AUTO_TEST_SUITE(RbmNetworkTest);
+
+template<typename MatType = arma::mat>
+void BuildVanillaNetwork(MatType& trainData,
+                         const size_t hiddenLayerSize)
+{
+  /*
+   * Construct a feed forward network with trainData.n_rows input nodes,
+   * hiddenLayerSize hidden nodes and trainLabels.n_rows output nodes. The
+   * network structure looks like:
+   *
+   *  Visible       Hidden        
+   *  Layer         Layer         
+   * +-----+       +-----+       
+   * |     |       |     |            
+   * |     +<----->|     |
+   * |     |       |     | 
+   * +-----+       +-----+     
+   *        
+   */
+  arma::mat output;
+  BinaryLayer<> visible(trainData.n_rows, hiddenLayerSize, 1);
+  BinaryLayer<> hidden(hiddenLayerSize, trainData.n_rows, 0);
+  GaussianInitialization gaussian(0, 0.1);
+  RBM<GaussianInitialization, BinaryLayer<>, BinaryLayer<>> model(trainData,
+      gaussian, visible, hidden, 1,  true);
+  model.Reset();
+  // Set the parmaeters from a learned rbm sklearn random state 23
+  model.Parameters() = arma::mat(
+      "-0.23224054, -0.23000632, -0.25701271, -0.25122418, -0.20716651,"
+      "-0.20962217, -0.59922456, -0.60003836, -0.6, -0.625, -0.475;");
+  // Check Weight Shared
+  BOOST_REQUIRE_CLOSE(arma::accu(model.VisibleLayer().Weight() -
+      model.HiddenLayer().Weight()), 0, 1e-14);
+
+  // Check free energy
+  arma::vec freeEnergy = arma::mat(
+      "-0.87523715, 0.50615066, 0.46923476, 1.21509084;");
+  arma::vec calcultedFreeEnergy(4);
+  calcultedFreeEnergy.zeros();
+  for (size_t i = 0; i < trainData.n_cols; i++)
+  {
+    model.VisibleLayer().ForwardPreActivation(std::move(trainData.col(i)),
+        std::move(output));
+    calcultedFreeEnergy(i) = model.FreeEnergy(std::move(trainData.col(i)));
+  }
+  for (size_t i = 0; i < freeEnergy.n_elem; i++)
+    BOOST_REQUIRE_CLOSE(calcultedFreeEnergy(i), freeEnergy(i), 1e-5);
+}
+BOOST_AUTO_TEST_CASE(MiscTest)
+{
+  /**
+   * Train and evaluate a vanilla network with the specified structure.
+   */
+
+  arma::mat X = arma::mat("0, 0, 0;"
+                          "0, 1, 1;"
+                          "1, 0, 1;"
+                          "1, 1, 1;");
+  X = X.t();
+  BuildVanillaNetwork<>(X, 2);
+}
+
+BOOST_AUTO_TEST_CASE(ClassificationTest)
+{
+  // Normalised dataset
+  int hiddenLayerSize = 100;
+  arma::mat trainData, testData, dataset;
+  arma::vec trainLabelsTemp, testLabelsTemp;
+  data::Load("mnisttrain.txt", trainData, true);
```

Comment:
I'm curious, have you tested to store the datasets as ``arma_binary``? I think it would be great if we could keep the size of the repo and datasets low.